File: 400/450 - Delete Node in a BST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        TreeNode left, right, temp;
        
        if (root == null)
            return null;
        
        left = root.left;
        right = root.right;
        
        if (root.val > key) {
            root.left = deleteNode(left, key);
            return root;
        }
        
        if (root.val < key) {
            root.right = deleteNode(right, key);
            return root;
        }

        if (left == null)
            return right;
        if (right == null)
            return left;

        temp = left;
        while (temp.right != null)
            temp = temp.right;
        
        temp.right = right.left;
        right.left = left;
        
        return right;
    }
}

File: csharp/450.delete-node-in-a-bst.cs
/*
 * @lc app=leetcode id=450 lang=csharp
 *
 * [450] Delete Node in a BST
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode DeleteNode(TreeNode root, int key) {
        if (root == null) {
            return null;
        }
        int val = root.val;
        if (key > val) {
            root.right = DeleteNode(root.right, key);
            return root;
        }
        if (key < val) {
            root.left = DeleteNode(root.left, key);
            return root;
        }
        if (root.left == null) {
            return root.right;
        }
        if (root.right == null) {
            return root.left;
        }
        TreeNode node = root.right;
        while (node.left != null) {
            node = node.left;
        }
        root.val = node.val;
        root.right = DeleteNode(root.right, root.val);
        return root;
    }
}

